modules/codegen/src/smithy4s/codegen/IR.scala

```diff
@@ -284,28 +302,54 @@ object TypedNode {
     }
   }
   object AltValueTN {
+    implicit val altValueTNTraverse: Traverse[AltValueTN] =
+      new Traverse[AltValueTN] {
+        def traverse[G[_]: Applicative, A, B](
+            fa: AltValueTN[A]
+        )(f: A => G[B]): G[AltValueTN[B]] =
+          fa match {
+            case ProductAltTN(value) => f(value).map(ProductAltTN(_))
+            case TypeAltTN(value)    => f(value).map(TypeAltTN(_))
+          }
+        def foldLeft[A, B](fa: AltValueTN[A], b: B)(f: (B, A) => B): B = ???
+        def foldRight[A, B](fa: AltValueTN[A], lb: Eval[B])(
+            f: (A, Eval[B]) => Eval[B]
+        ): Eval[B] = ???
+      }
+
     case class ProductAltTN[A](value: A) extends AltValueTN[A]
     case class TypeAltTN[A](value: A) extends AltValueTN[A]
   }
 
-  implicit val typedNodeFunctor: Functor[TypedNode] = new Functor[TypedNode] {
-    def map[A, B](fa: TypedNode[A])(f: A => B): TypedNode[B] = fa match {
-      case EnumerationTN(ref, value, ordinal, name) =>
-        EnumerationTN(ref, value, ordinal, name)
-      case StructureTN(ref, fields) =>
-        StructureTN(ref, fields.map(_.map(_.map(f))))
-      case NewTypeTN(ref, target) =>
-        NewTypeTN(ref, f(target))
-      case AltTN(ref, altName, alt) =>
-        AltTN(ref, altName, alt.map(f))
-      case MapTN(values) =>
-        MapTN(values.map(_.leftMap(f).map(f)))
-      case CollectionTN(collectionType, values) =>
-        CollectionTN(collectionType, values.map(f))
-      case PrimitiveTN(prim, value) =>
-        PrimitiveTN(prim, value)
+  implicit val typedNodeTraverse: Traverse[TypedNode] =
+    new Traverse[TypedNode] {
+      def traverse[G[_], A, B](
+          fa: TypedNode[A]
+      )(f: A => G[B])(implicit F: Applicative[G]): G[TypedNode[B]] = fa match {
+        case EnumerationTN(ref, value, ordinal, name) =>
+          F.pure(EnumerationTN(ref, value, ordinal, name))
+        case StructureTN(ref, fields) =>
+          fields.traverse(_.traverse(_.traverse(f))).map(StructureTN(ref, _))
+        case NewTypeTN(ref, target) =>
+          f(target).map(NewTypeTN(ref, _))
+        case AltTN(ref, altName, alt) =>
+          alt.traverse(f).map(AltTN(ref, altName, _))
+        case MapTN(values) =>
+          values
+            .traverse { case (k, v) =>
+              (f(k), f(v)).tupled
+            }
+            .map(MapTN(_))
+        case CollectionTN(collectionType, values) =>
+          values.traverse(f).map(CollectionTN(collectionType, _))
+        case PrimitiveTN(prim, value) =>
+          F.pure(PrimitiveTN(prim, value))
+      }
+      def foldLeft[A, B](fa: TypedNode[A], b: B)(f: (B, A) => B): B = ???
```

Comment:
I'm pretty sure `foldLeft` can be derived from traverse (even mechanically, using `Const`) - it should be possible to implement, then. As a principle we should also test these instances

Reply:
> is it because they can't be implement or because you did not need them? because it sure looks like a trap to me!

honestly, it's because it's tedious to implement, and it's only used in a very specific context that only requires traverse. See [this issue](https://github.com/typelevel/cats/issues/2975) 
 
> I'm pretty sure foldLeft can be derived from traverse (even mechanically, using Const) - it should be possible to implement, then. As a principle we should also test these instances

it may be the case for foldLeft, but I think I've never seen a generic implementation for foldRight that was lawful (something to do with stack safety iirc), so I'm not even sure it's possible. I've certainly [given it a go myself](https://gist.github.com/Baccata/0b972953cd65421d6a475b37aadad65b), and I can't even remember if it worked.

Reply:
> As a principle we should also test these instances 

meh, it's not code that's used at runtime and I have to much to do ...